# Data Validation Report

Runs the reusable `validate_dataset` utility against all three raw datasets and prints a summary of findings.

In [2]:
import sys
import os
import pandas as pd

# Make src/ importable from the notebooks/ directory
sys.path.append(os.path.join(os.path.dirname(os.getcwd()), 'src'))
# When running from notebooks/ folder:
sys.path.append('../src')

from validator import validate_dataset

print('validator imported successfully')

ImportError: cannot import name 'validate_dataset' from 'validator' (c:\Users\mtawa\AppData\Local\Programs\Python\Python312\Lib\site-packages\validator\__init__.py)

## 1 — Flight Data Validation

In [2]:
validate_dataset(
    file_path='../data/raw/flight_data_2024.csv',
    output_folder='../outputs/reports/validation_flight',
    target_column=None,          # target not yet created at raw stage
    id_columns=None,
    missing_threshold=30,
    outlier_method='iqr'
)

d:\uni\Semester 10\Data Science\Project\flight-delay-predictor\src\validator.py:35: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, na_values=missing_placeholders)


Validation reports saved in folder: ../outputs/reports/validation_flight


In [3]:
# Print the human-readable summary
with open('../outputs/reports/validation_flight/summary_report.txt', 'r') as f:
    print(f.read())

DATA VALIDATION SUMMARY
Dataset file: ../data/raw/flight_data_2024.csv
Rows: 7079081
Columns: 35
Duplicate rows: 0
Columns with >30% missing: cancellation_code
Columns with one unique value: year
Columns with leading/trailing spaces: None
Columns with inconsistent casing: None
Numeric columns with outliers: dep_delay, arr_delay, taxi_in, distance, crs_elapsed_time, taxi_out, air_time, actual_elapsed_time, op_carrier_fl_num
Class imbalance: Not checked


In [4]:
# Show top missing columns
missing_df = pd.read_csv('../outputs/reports/validation_flight/missing_values_report.csv')
print('=== Missing Values (Flight Data) ===')
print(missing_df[missing_df['Missing Count'] > 0].to_string(index=False))

=== Missing Values (Flight Data) ===
             Column  Missing Count  Missing %
  cancellation_code        6982766      98.64
actual_elapsed_time         113814       1.61
          arr_delay         113814       1.61
           air_time         113814       1.61
            taxi_in          97856       1.38
          wheels_on          97856       1.38
           arr_time          97854       1.38
           taxi_out          95734       1.35
         wheels_off          95734       1.35
           dep_time          92659       1.31
          dep_delay          92970       1.31
   crs_elapsed_time              1       0.00
  op_carrier_fl_num              1       0.00


In [5]:
# Show quality issues
issues_df = pd.read_csv('../outputs/reports/validation_flight/quality_issues_report.csv')
print('=== Quality Issues (Flight Data) ===')
print(issues_df.to_string(index=False))

=== Quality Issues (Flight Data) ===
                          Issue Type                                                                                                               Columns
       Columns with one unique value                                                                                                                  year
           Columns with >30% missing                                                                                                     cancellation_code
Columns with leading/trailing spaces                                                                                                                   NaN
    Columns with inconsistent casing                                                                                                                   NaN
       Numeric columns with outliers dep_delay, arr_delay, taxi_in, distance, crs_elapsed_time, taxi_out, air_time, actual_elapsed_time, op_carrier_fl_num


In [6]:
# Show outlier report
outlier_df = pd.read_csv('../outputs/reports/validation_flight/outlier_report.csv')
print('=== Outliers (Flight Data) ===')
print(outlier_df[outlier_df['Outlier Count'] > 0].to_string(index=False))

=== Outliers (Flight Data) ===
             Column  Outlier Count  Outlier %
          dep_delay         909066      12.84
          arr_delay         670540       9.47
            taxi_in         466184       6.59
           distance         420064       5.93
   crs_elapsed_time         360026       5.09
           taxi_out         355622       5.02
           air_time         349654       4.94
actual_elapsed_time         341997       4.83
  op_carrier_fl_num            888       0.01


## 2 — Weather Data Validation

In [ ]:
validate_dataset(
    file_path='../data/raw/weather_data_real.csv',
    output_folder='../outputs/reports/validation_weather',
    target_column=None,
    id_columns=None,
    missing_threshold=30,
    outlier_method='iqr'
)

In [8]:
with open('../outputs/reports/validation_weather/summary_report.txt', 'r') as f:
    print(f.read())

DATA VALIDATION SUMMARY
Dataset file: ../data/raw/weather_data.csv
Rows: 1000000
Columns: 6
Duplicate rows: 0
Columns with >30% missing: None
Columns with one unique value: None
Columns with leading/trailing spaces: None
Columns with inconsistent casing: None
Numeric columns with outliers: None
Class imbalance: Not checked


In [9]:
issues_df = pd.read_csv('../outputs/reports/validation_weather/quality_issues_report.csv')
print('=== Quality Issues (Weather Data) ===')
print(issues_df.to_string(index=False))

=== Quality Issues (Weather Data) ===
                          Issue Type  Columns
       Columns with one unique value      NaN
           Columns with >30% missing      NaN
Columns with leading/trailing spaces      NaN
    Columns with inconsistent casing      NaN
       Numeric columns with outliers      NaN


## 3 — Global Airports Data Validation

In [10]:
validate_dataset(
    file_path='../data/raw/global_airports.csv',
    output_folder='../outputs/reports/validation_airports',
    target_column=None,
    id_columns=['IATA'],          # IATA should be unique per airport
    missing_threshold=30,
    outlier_method='iqr'
)

Validation reports saved in folder: ../outputs/reports/validation_airports


In [11]:
with open('../outputs/reports/validation_airports/summary_report.txt', 'r') as f:
    print(f.read())

DATA VALIDATION SUMMARY
Dataset file: ../data/raw/global_airports.csv
Rows: 6393
Columns: 13
Duplicate rows: 0
Columns with >30% missing: None
Columns with one unique value: None
Columns with leading/trailing spaces: AirportName, ICAO, City_Name, City_IATA, Country_CodeA3
Columns with inconsistent casing: AirportName, City_Name
Numeric columns with outliers: None
Class imbalance: Not checked


In [12]:
missing_df = pd.read_csv('../outputs/reports/validation_airports/missing_values_report.csv')
print('=== Missing Values (Airports Data) ===')
print(missing_df[missing_df['Missing Count'] > 0].to_string(index=False))

=== Missing Values (Airports Data) ===
            Column  Missing Count  Missing %
              ICAO           1302      20.37
          TimeZone             88       1.38
  UTC_Offset_Hours             88       1.38
UTC_Offset_Seconds             88       1.38
         City_IATA             62       0.97
              IATA             21       0.33
    Country_CodeA2             14       0.22
       AirportName              1       0.02


In [13]:
issues_df = pd.read_csv('../outputs/reports/validation_airports/quality_issues_report.csv')
print('=== Quality Issues (Airports Data) ===')
print(issues_df.to_string(index=False))

=== Quality Issues (Airports Data) ===
                          Issue Type                                                 Columns
       Columns with one unique value                                                     NaN
           Columns with >30% missing                                                     NaN
Columns with leading/trailing spaces AirportName, ICAO, City_Name, City_IATA, Country_CodeA3
    Columns with inconsistent casing                                  AirportName, City_Name
       Numeric columns with outliers                                                     NaN


## 4 — Fix Identified Issues in Airports Dataset

Strip leading/trailing whitespace and normalise casing for `AirportName` and `City_Name`.

In [14]:
airports = pd.read_csv('../data/raw/global_airports.csv')

str_cols = airports.select_dtypes(include='object').columns

print('Before cleaning:')
print(f'  Rows: {len(airports)}')
print(f'  Sample AirportName values: {airports["AirportName"].head(5).tolist()}')

# Strip whitespace on all string columns
for col in str_cols:
    airports[col] = airports[col].str.strip()

# Title-case AirportName and City_Name for consistency
airports['AirportName'] = airports['AirportName'].str.title()
airports['City_Name']   = airports['City_Name'].str.title()

# Upper-case IATA and ICAO codes
airports['IATA'] = airports['IATA'].str.upper()
airports['ICAO'] = airports['ICAO'].str.upper()

print('\nAfter cleaning:')
print(f'  Sample AirportName values: {airports["AirportName"].head(5).tolist()}')
print(f'  Remaining whitespace issues: {airports[str_cols].apply(lambda c: c.str.startswith(" ") | c.str.endswith(" ")).any().any()}')

# Save cleaned airports
os.makedirs('../data/processed', exist_ok=True)
airports.to_csv('../data/processed/airports_clean.csv', index=False)
print('\nSaved cleaned airports to data/processed/airports_clean.csv')

Before cleaning:
  Rows: 6393
  Sample AirportName values: ['Pilanesberg Intl', 'Clovis Muni', 'Cannon Afb', 'Scammon Bay Airport', 'Kapit               ']

After cleaning:
  Sample AirportName values: ['Pilanesberg Intl', 'Clovis Muni', 'Cannon Afb', 'Scammon Bay Airport', 'Kapit']
  Remaining whitespace issues: False

Saved cleaned airports to data/processed/airports_clean.csv


## 5 — Validation Summary

| Dataset | Rows | Key Issues Found |
|---------|------|------------------|
| Flight Data | 7,079,081 | `cancellation_code` 98.64% missing; ~1.3% missing in operational columns |
| Weather Data | 1,000,000 | No missing values; only 10 US cities covered |
| Airports Data | 6,393 | 20 missing IATA codes; whitespace & casing in names |

**Actions taken:**
- `cancellation_code` → dropped in `flight_data_pipeline.ipynb` (98.64% missing)
- Cancelled & diverted flights → filtered out (no arrival delay to predict)
- Airports whitespace & casing → fixed above and saved to `data/processed/airports_clean.csv`
- Full validation CSVs → saved under `outputs/reports/validation_*/`